EC-84: Revise Cache, Catalog and Checkout sections (#3)
ezsystems/commerce-documentation: docs/guide/catalog/catalog_cookbook/setting_up_variants_from_external_source.md

You can also set up variants if you are not using the default Content Types
and don't want to set up variants manually in the Back Office. 

The main goal is to map your variant structure to the [`VariantProductNode`](../catalog_api/variantproductsnode_and_orderablevariantnode/variantproductnode_and_orderablevariantnode.md) structure.

## How to start

!!! tip
    
    The main job of every CatalogFactory is to fill the [`VariantProductNode`](../catalog_api/variantproductsnode_and_orderablevariantnode/variantproductnode_and_orderablevariantnode.md) attribute [`$variantCharacteristics`](../catalog_api/variantproductsnode_and_orderablevariantnode/simplevariantcharacteristics.md). The basic concept is to take information from provided `dataMap` (e.g. `dataMap` is coming from the data provider, either from the content model or from an external source) and create an attribute `$variantCharacteristic` for the `VariantProductNode`.

> Example for `Ez5CatalogFactory`: `dataMap` from th data provider

Reviewer: ```suggestion
Example for `Ez5CatalogFactory`: `dataMap` from the data provider
```

![](../../img/catalog_cookbook_2.png)

Example for `Ez5CatalogFactory`: the information was already mapped in the `$variantCharacteristics` attribute in catalog element: 

![](../../img/catalog_cookbook_3.png)

## Mapping variant information

### 1. Create data provider and catalog factory

The first thing you need to do to create variants is to write your own data provider and factory for catalog elements. 

!!! tip

    It is good practice to define a configuration for templates.
    Depending on which type of catalog element the provider/factory creates, different template should be used.
    This is the default configuration:

    ```
    silver_eshop.default.catalog_template.CatalogNode: Catalog:catalog.html.twig
    silver_eshop.default.catalog_template.OrderableProductNode: Catalog:product.html.twig
    silver_eshop.default.catalog_template.VariantProductNode: Catalog:product_variants.html.twig
    ```

>     You can find more information about this is [Catalog UI](../catalog_ui/catalog_ui.md).

Reviewer: ```suggestion
    You can find more information about this in [Catalog UI](../catalog_ui/catalog_ui.md).
```

If you are writing your own provider and factory from scratch, you have to implement all necessary methods.
This example only shows methods related to variants.

### 2. Implement a method for creating of VariantProductNode

#### Determine which method to call and when

The starting point is the `createCatalogElement()` method.
This is the place where the factory decides which function it internally uses to create the catalog element.

!!! tip

>     It is good practice to define a configuration which method should be used depending on e.g. the class identifier (Content Type).

Reviewer: ```suggestion
    It is good practice to configure which method should be used depending on e.g. the class identifier (Content Type).
```

    This is the default configuration:

    ```
    silver_eshop.default.catalog_factory.ses_category: createCatalogNode
    silver_eshop.default.catalog_factory.ses_product: createOrderableProductNode
    ```

> Because in this example (standard implementation), the variants are stored as Content items of the same Content Type as regular products (`ses_product`),

Reviewer: ```suggestion
`createCatalogElement()` must do additional differentiation when to call the method `createOrderableProductNode()` and when `createVariantProductNode()`
because in the below example (standard implementation), the variants are stored as Content items of the same Content Type as regular products (`ses_product`),.
```

> `createCatalogElement()` must do additional differentiation when to call the method `createOrderableProductNode()` and when `createVariantProductNode()` (see below).

Reviewer: ```suggestion
```

> In the default implementation the factory calls the `createVariantProductNode()` method whenever [EzMatrix with variants](../catalog_features/product_variants/product_variants.md) is filled.

Reviewer: ```suggestion
In the default implementation the factory calls the `createVariantProductNode()` method whenever [eZ Matrix with variants](../catalog_features/product_variants/product_variants.md) is filled.
```

> You can adapt this behavior. For example, you can receive a special flag from ERP that tells you if the product is a variant.

Reviewer: ```suggestion
You can adapt this behavior. For example, you can receive a special flag from ERP that informs you if the product is a variant.
```

    } 
    ```

#### Implement `createVariantProductNode()`

Next, you need to implement the `createVariantProductNode()` method from the abstract class `CatalogFactory`:

``` 
abstract public function createVariantProductNode($rawData = null);
```

> Example from `Ez5CatalogFactory` (we use information stored in ez matrix (`ses_variants`) which is located in `dataMap` (provided by the data provider).

Reviewer: ```suggestion
Example from `Ez5CatalogFactory` uses information stored in eZ Matrix (`ses_variants`) which is located in `dataMap` (provided by the data provider).
```

!!! tip

    `$this->extractVariants()` extracts the variant information from the `dataMap` and returns an array with a single element
    that is a [`SimpleVariantCharacteristics` object](../catalog_api/variantproductsnode_and_orderablevariantnode/simplevariantcharacteristics.md).

### 3. Fill VariantProductNode attribute $variantCharacteristics

In your factory you need to map your `dataMap` information into the `VariantProductNode` attribute [`$variantCharacteristics`](../catalog_api/variantproductsnode_and_orderablevariantnode/simplevariantcharacteristics.md). The structure is described below.

> All information about variants is stored in this attribute (including prices and pictures). It contains three pieces of information:

Reviewer: ```suggestion
All information about variants is stored in this attribute (including prices and pictures). It contains the following pieces of information:
```

- `characteristics` contains all information about characteristics (selectable options) like images or labels for displaying options in templates
- `variantCodes` is a set of all possible combinations of variant characteristics (variants). It contains a list of the respective variant codes, which store a particular combination of the characteristics. The record structure is:

```
- variantCode =>
    - characteristicIdentifier => characteristicCode
    - ...
- ...
```

> In this example, the `characteristicIdentifier` is the index number of the variant level

Reviewer: ```suggestion
In the above example, the `characteristicIdentifier` is the index number of the variant level
```

which is provided by the structure of eZ Matrix.

> - `variantAttributes` - contains all additional information for each variant (country, price, labels etc). Attributes that are displayed in B2B need to follow special rules:

Reviewer: ```suggestion
- `variantAttributes` - contains all additional information for each variant (country, price, labels etc.). Attributes that are displayed in B2B need to follow special rules:
```
